java-checks/src/main/java/org/sonar/java/checks/security/ControllingPermissionsCheck.java

```diff
@@ -0,0 +1,141 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.security;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.MethodMatcherCollection;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+
+@Rule(key = "S4834")
+public class ControllingPermissionsCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String MESSAGE = "Make sure that Permissions are controlled safely here.";
+
+  private static final List<String> ANNOTATIONS = Collections.unmodifiableList(Arrays.asList(
+    "org.springframework.security.access.prepost.PostAuthorize",
+    "org.springframework.security.access.prepost.PostFilter",
+    "org.springframework.security.access.prepost.PreAuthorize",
+    "org.springframework.security.access.prepost.PreFilter",
+    "org.springframework.security.access.annotation.Secured"));
+
+  private static final List<String> JSR_250_ANNOTATIONS = Collections.unmodifiableList(Arrays.asList(
+    "javax.annotation.security.RolesAllowed",
+    "javax.annotation.security.PermitAll",
+    "javax.annotation.security.DenyAll"));
+
+  private static final List<String> INTERFACES = Collections.unmodifiableList(Arrays.asList(
+    "org.springframework.security.access.AccessDecisionVoter",
+    "org.springframework.security.access.AccessDecisionManager",
+    "org.springframework.security.access.AfterInvocationProvider",
+    "org.springframework.security.access.PermissionEvaluator",
+    "org.springframework.security.access.expression.SecurityExpressionOperations",
+    "org.springframework.security.access.expression.method.MethodSecurityExpressionHandler",
+    "org.springframework.security.core.GrantedAuthority",
+    "org.springframework.security.acls.model.PermissionGrantingStrategy"));
+
+  private static final List<String> CLASSES = Collections.singletonList("org.springframework.security.config.annotation.method.configuration.GlobalMethodSecurityConfiguration");
```

Comment:
Naming of this constant can probably be improved to designate which CLASSES/INTERFACES/ANNOTATIONS it targets.

Reply:
Given the usage of those list I'm wondering if it would not be better to have the predicates as constants (so you can eventually compose them when needed) rather than storing the lists directly.